src/signature.js

```diff
@@ -152,13 +152,40 @@ const substituteTypes = argTypes => type => R.compose(
   R.prop('keys'),
 )(type);
 
+//  :: Type -> (Type -> Type)
+const fromUnaryType = t => $.UnaryType(t.name, t.url, t._test, t.types.$1.extractor);
+//  :: Type -> (Type -> Type -> Type)
+const fromBinaryType = t => $.BinaryType(t.name, t.url, t._test, t.types.$1.extractor, t.types.$2.extractor);
+
+// :: Type -> Boolean
+const isUnaryType = R.propEq('type', 'UNARY');
+// :: Type -> Boolean
+const isBinaryType = R.propEq('type', 'BINARY');
+
+// :: [Type] -> Type -> Type
+const constructNewTypeFromType = argTypes => type => R.compose(
+  R.apply(R.__, argTypes),
+  R.cond([
+    [isUnaryType, fromUnaryType],
+    [isBinaryType, fromBinaryType],
+    [R.T, (x) => {
+      // Actually it won't happen.
+      // This is to ensure a clear error if one day sactuary-def change constants or something else.
+      throw new TypeError(
+        `Type ${type.name} should be recreated with Types: ${R.map(R.prop('name'), argTypes)} ` +
+        `but it haven't got a proper function recreator for type ${type.type}.`
+      );
+    }]
+  ])
+)(type);
```

Comment:
The structure of this function seems convoluted to me. I suggest something like this:

```javascript
//    constructNewTypeFromType :: Array Type -> Type -> Type
const constructNewTypeFromType = ts => t => {
  switch (t.type) {
    case 'UNARY':   return fromUnaryType(t)(ts[0]);
    case 'BINARY':  return fromBinaryType(t)(ts[0], ts[1]);
    default:        throw new TypeError('...');
  }
};
```

This function could be further simplified by combining it with `constructType`.